src/Processors/Formats/Impl/BSONEachRowRowInputFormat.h

```diff
@@ -0,0 +1,113 @@
+ * \x0E symbol                                 | String/FixedString
+ * \x10 int32                                  | Int32/Decimal32
+ * \x12 int64                                  | Int64/Decimal64/DateTime64
+ * \x11 uint64                                 | UInt64
+ *
+ * Other BSON types are not supported.
+ * Also, we perform conversion between different integer types
+ * (for example, you can insert BSON int32 value into ClickHouse UInt8)
+ * Big integers and decimals Int128/UInt128/Int256/UInt256/Decimal128/Decimal256
+ * can be parsed from BSON Binary value with \x00 binary subtype. In this case
+ * we validate that the size of binary data equals the size of expected value.
+ *
+ * Note: this format will not work on Big-Endian platforms.
+ */
+
+class ReadBuffer;
+class BSONEachRowRowInputFormat final : public IRowInputFormat
+{
+public:
+    BSONEachRowRowInputFormat(
+        ReadBuffer & in_, const Block & header_, Params params_, const FormatSettings & format_settings_);
+
+    String getName() const override { return "BSONEachRowRowInputFormat"; }
+    void resetParser() override { }
+
+private:
+    void readPrefix() override { }
+    void readSuffix() override { }
+
+    bool readRow(MutableColumns & columns, RowReadExtension & ext) override;
+    bool allowSyncAfterError() const override { return true; }
+    void syncAfterError() override;
+
+    size_t columnIndex(const StringRef & name, size_t key_index);
+
+    using ColumnReader = std::function<void(StringRef name, BSONType type)>;
+
+    bool readField(IColumn & column, const DataTypePtr & data_type, BSONType bson_type);
+    void skipUnknownField(BSONType type, const String & key_name);
+
+    void readTuple(IColumn & column, const DataTypePtr & data_type, BSONType bson_type);
+    void readArray(IColumn & column, const DataTypePtr & data_type, BSONType bson_type);
+    void readMap(IColumn & column, const DataTypePtr & data_type, BSONType bson_type);
+
+    const FormatSettings format_settings;
+
+    /// Buffer for the read from the stream field name. Used when you have to copy it.
+    String current_key_name;
+
+    /// Set of columns for which the values were read. The rest will be filled with default values.
+    std::vector<UInt8> read_columns;
+    /// Set of columns which already met in row. Exception is thrown if there are more than one column with the same name.
+    std::vector<UInt8> seen_columns;
+    /// These sets may be different, because if null_as_default=1 read_columns[i] will be false and seen_columns[i] will be true
+    /// for row like {..., "non-nullable column name" : null, ...}
+
+    /// Hash table match `field name -> position in the block`.
+    Block::NameMap name_map;
+
+    /// Cached search results for previous row (keyed as index in JSON object) - used as a hint.
+    std::vector<Block::NameMap::LookupResult> prev_positions;
+
+    DataTypes types;
+
+    size_t current_document_start;
+    BSONSizeT current_document_size;
```

Comment:
In the current implementation, the maximal document size is 4GB. So the user can provide a malformed document with intmax in the header, we will allocate and destroy 4GB of memory, and also it's allocated under MemoryTracket, so it shouldn't be any vulnerability here.

But mongo [has](https://www.mongodb.com/docs/manual/reference/limits/#mongodb-limit-BSON-Document-Size) 16MB limit for the size of a document. Probably we also can do the same. Is it needed, what do you think?

Reply:
I think we can do similar to JSON and restrict max document size only for parallel parsing to avoid allocating big memory chunks (without parallel parsing allocated memory size doesn't depend on document size)